Add the AWS guide (#610)
elastic/observability-docs: docs/en/observability/monitor-aws.asciidoc

Next, you are going to configure the {filebeat} output to ship data to {ess}.

Use the {filebeat} keystore to store {filebeat-ref}/keystore.html[secure
settings]. Store the Cloud ID in the keystore.
[source,bash]
----
./filebeat keystore create
echo -n "<Your Deployment Cloud ID>" | ./filebeat keystore add CLOUD_ID --stdin
----

To store logs in {es} with minimal permissions, create an API key to send
data from {filebeat} to {ess}. Log into {kib} (you can do so from the Cloud
Console without typing in any permissions) and select *Management* -> *Dev
Tools*. Send the following request:

[source,console]
----
POST /_security/api_key
{
  "name": "filebeat-monitor-gcp",
  "role_descriptors": {
    "filebeat_writer": {
      "cluster": [
        "monitor",
        "read_ilm",
        "cluster:admin/ingest/pipeline/get", <1>
        "cluster:admin/ingest/pipeline/put" <1>
      ],
      "index": [
        {
          "names": ["filebeat-*"],
          "privileges": ["view_index_metadata", "create_doc"]
        }
      ]
    }
  }
}
----

<1> {filebeat} needs extra cluster permissions to publish logs, which differs
from the {metricbeat} configuration used previously. You can find more

> details {filebeat-ref}/feature-roles.html[here].

Author: I removed it.

`cloudwatch`, `elb` and `s3access`. In this tutorial we are going to show you a
few examples using the `ec2` and the `s3access` filesets. The `ec2` fileset is
used to ship and process logs stored in cloudwatch and exported them to a S3
bucket. The `s3access` fileset is used when S3 access logs need to be collected,
it provides detailed records for the requests that are made to a bucket. Server
access logs are useful for many applications. For example, access log
information can be useful in security and access audits. It can also help you
learn about your customer base and understand your Amazon S3 bill.

Let's enable the {aws} module in {metricbeat}.

[source,bash]
----
./filebeat modules enable aws
----

Edit the `modules.d/aws.yml` file with the following configurations.

[source,yml]
----
- module: aws
  cloudtrail:
    enabled: false <1>
  cloudwatch:
    enabled: false <1>
  ec2:
    enabled: true <2>
    var.credential_profile_name: fb-aws <3>
    var.queue_url: https://sqs.eu-central-1.amazonaws.com/836370109380/howtoguide-tutorial <4>
  elb:
    enabled: false <1>
  s3access:
    enabled: false <1>
  vpcflow:
    enabled: false <1>
----

> <1> Disables the filesets.

Author: I removed the ones

<2> Enables the `ec2` fileset.
<3> This is the aws profile define following the https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-files.html[AWS standard].
<4> Add the URL to the queue containing notifications around the bucket containing the EC2 logs

You can now upload your logs to the S3 bucket. If you are using cloudwatch,
make sure to edit the policy of your bucket like shown in the https://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/S3ExportTasksConsole.html[step 3 of the {aws} user guide]
it will help you avoiding permissions issues.

Start {filebeat} to collect the logs.

[source,bash]
----
./filebeat -e
----

> Let's now configure another fileset and we are going to use the `s3access`
> one. The goal here is to be able to monitor how people access the bucket we
> created. This is what we achieved:
> 
> image::one-bucket-archi.png[Current Architecture]
> 
> Now what we need to do is to create another bucket and another queue. In order
> to create the following architecture:
> 
> image::two-buckets-archi.png[Architecture with Access Logging Enabled]

Reviewer: It seems like we're switching gears, but then we go back to what we've already achieved. What if we show the first screenshot before starting to switch context?

```suggestion
Here's what we've achieved so far:

image::one-bucket-archi.png[Current Architecture]

Now, let's configure the `s3access` fileset.
The goal here is to be able to monitor how people access the bucket we
created. To do this, we'll create another bucket and another queue.
The new architecture will look like this:

image::two-buckets-archi.png[Architecture with Access Logging Enabled]
```

> Make sure the event notifications on the bucket you created is enabled and is
> sending notifications to the new queue you also just created. Now you need to go
> back to the first bucket you created, go to the properties, go to **Server access
> logging**, specify that you want to ship the access logs to the bucket you last
> created.

Reviewer: ```suggestion
Create a new S3 bucket and SQS queue. Ensure that the event notifications on the new bucket are enabled,
and that it's sending notifications to the new queue.

Now go back to the first bucket, and go to *Properties* > *Server access
logging*. Specify that you want to ship the access logs to the bucket you most recently
created.
```

image::Server-Access-Logging.png[Enabling Server Access Logging]

Copy the URL of the queue you created. Edit the `modules.d/aws.yml`file with
the following configurations.

[source,yml]
----
- module: aws
  cloudtrail:
    enabled: false <1>
  cloudwatch:
    enabled: false <1>
  ec2:
    enabled: true <2>
    var.credential_profile_name: fb-aws <3>
    var.queue_url: https://sqs.eu-central-1.amazonaws.com/836370109380/howtoguide-tutorial <4>
  elb:
    enabled: false <1>
  s3access:
    enabled: true <5>
    var.credential_profile_name: fb-aws <3>
    var.queue_url: https://sqs.eu-central-1.amazonaws.com/836370109380/access-log <5>
  vpcflow:
    enabled: false <1>
----

> <1> Disables the filesets.

Reviewer: Similar thought here in regards to removing the <1>s

Author: I removed the ones

<2> Enables the `ec2` fileset.
<3> This is the aws profile define following the https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-files.html[AWS standard].
<4> Add the URL to the queue containing notifications around the bucket containing the EC2 logs
<5> Add the URL to the queue containing notifications around the bucket containing the S3 access logs

Once you have edited the config file, you need to restart filebeat. To stop
flebeat, you can press CTRL + C in the terminal. Now let's restart filebeat by
running the following command:

[source,bash]
----
./filebeat -e
----

[discrete]
==== Step 6: Visualizing Logs

Now that the logs are being shipped to Elasticsearch we can visualize them in

> Kibana. Go to Kibana, in the interface click on the left panel, click on
> **Logs** to see how your raw logs look like:

Reviewer: Changing this to be consistent with other Kibana terminology

```suggestion
Kibana. To see the raw logs, open the main menu in Kibana, then click
**Logs**:
```

image::EC2-logs.png[EC2 logs in the Logs UI]

> The filesets we used in the previous steps also comes with prebuilt dashboard

Reviewer: ```suggestion
The filesets we used in the previous steps also come with prebuilt dashboards
```

> that you can use to visualize the data. In Kibana, on the left panel, click on
> **dashboard**, search for S3 and select the dashboard called:

Reviewer: ```suggestion
that you can use to visualize the data. In Kibana, open the main menu and click
**Dashboard**. Search for S3 and select the dashboard called:
```

**[Filebeat AWS] S3 Server Access Log Overview**:

image::S3-Server-Access-Logs.png[S3 Server Access Log Overview]

> This give you an overview of how your S3 buckets are being accessed.

Reviewer: ```suggestion
This gives you an overview of how your S3 buckets are being accessed.
```

[discrete]
==== Step 7: Collecting Metrics

To monitor your {aws} infrastructure you will need to first make sure your
infrastructure data are being shipped to CloudWatch. To ship the data to
Elasticsearch we are going to use the {aws} module from Metricbeat. This module
periodically fetches monitoring metrics from AWS CloudWatch using
**GetMetricData** API for {aws} services.

[IMPORTANT]
====
Extra AWS charges on CloudWatch API requests will be generated by this module.
Please see https://www.elastic.co/guide/en/beats/metricbeat/current/metricbeat-module-aws.html#aws-api-requests[AWS API requests] for more details.
====

[discrete]
==== Step 8: Install {metricbeat}

On a new terminal window, download and install {metricbeat}.

include::{beats-repo-dir}/tab-widgets/install-widget-metricbeat.asciidoc[]

[discrete]
==== Set up the assets

> Similarly to {filebeat}, {metricbeat} comes with predefined assets for parsing,

Reviewer: ```suggestion
Similar to {filebeat}, {metricbeat} comes with predefined assets for parsing,
```

        }
      ]
    }
  }
}
----

. The response contains an `api_key` and an `id` field, which can be stored in
the {metricbeat} keystore in the following format: `id:api_key`.
+
[source,bash]
----
echo -n "IhrJJHMB4JmIUAPLuM35:1GbfxhkMT8COBB4JWY3pvQ" | ./metricbeat keystore add ES_API_KEY --stdin
----
+
[NOTE]
=====
Make sure you specify the `-n` parameter; otherwise, you will have
painful debugging sessions due to adding a newline at the end of
your API key.
=====

. To see if both settings have been stored, run the following command:
+
[source,bash]
----
./metricbeat keystore list
----

. To configure {metricbeat} to output to {ess}, edit the `metricbeat.yml`
configuration file. Add the following lines to the end of the file.
+
[source,yml]
----
cloud.id: ${CLOUD_ID}
output.elasticsearch:
  api_key: ${ES_API_KEY}
----

. Finally, test if the configuration is working. If it is not working,
verify if you used the right credentials and add them again.
+
[source,bash]
----
./metricbeat test output
----

Now that the output is working, you are going to set up the {aws} module.

[discrete]
==== Step 9: Configuring the {aws} module

> To collect the metrics from your {aws} infrastructure we will use the
> {metricbeat-ref}/metricbeat-module-aws.html[{aws}] module. The {aws} module is
> containing a lot of metricsets: `billing`, `cloudwatch`, `dynamodb`, `ebs`,
> `ec2`, `elb`, `lambda` and many more. Each metricsets is created to help you
> stream and process your data. In this tutorial we are going to show you a
> few examples using the `ec2` and the `billing` metricsets.

Reviewer: ```suggestion
To collect metrics from your {aws} infrastructure, we'll use the
{metricbeat-ref}/metricbeat-module-aws.html[{metricbeat} {aws}] module. This module
contains many metricsets: `billing`, `cloudwatch`, `dynamodb`, `ebs`,
`ec2`, `elb`, `lambda`, and many more. Each metricset is created to help you
stream and process your data. In this tutorial, we're going to show you a
few examples using the `ec2` and the `billing` metricsets.
```

    group_by_dimension_keys:
      - "AZ"
      - "INSTANCE_TYPE"
      - "SERVICE"
    group_by_tag_keys:
      - "aws:createdBy"
- module: aws <1>
  period: 300s <2>
  metricsets:
    - ec2 <3>
  credential_profile_name: mb-aws <4>
----
+
<1> Defines the module that is going to be used.
<2> Defines the period at which the metrics are going to be collected
<3> Defines the metricset that is going to be used.
<4> This is the aws profile define following the https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-files.html[AWS standard].

Make sure that the {aws} user used to collect the metrics from CloudWatch has at
least the following permissions attached to it:

[source,yml]
----
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Action": [
                "ec2:DescribeInstances",
                "ec2:DescribeRegions",
                "cloudwatch:GetMetricData",
                "cloudwatch:ListMetrics",
                "sts:GetCallerIdentity",
                "iam:ListAccountAliases",
                "tag:getResources",
                "ce:GetCostAndUsage"
            ],
            "Effect": "Allow",
            "Resource": "*"
        }
    ]
}

> ----

Reviewer: Oh snap! Where is this for Filebeat? This is exactly what I needed above

Author: I found the following minimal permission:

```
s3:GetObject
sqs:ReceiveMessage
sqs:ChangeMessageVisibility
sqs:DeleteMessage
```

You can now start Metricbeat:

[source,bash]
----
./metricbeat -e
----

[discrete]
==== Step 10: Visualizing Metrics

Now that the metrics are being streamed to Elasticsearch we can visualize them in

> Kibana. Go to Kibana, in the interface click on the left panel, click on
> **Metrics** to see how your infrastructure look like. Once in the **Metrics UI**
> make sure to show the **{aws}** source and the **EC2 Instances**:

Reviewer: ```suggestion
Kibana. In Kibana, open the main menu and click
**Metrics**. Make sure to show the **{aws}** source and the **EC2 Instances**:
```

image::EC2-instances.png[Your EC2 Infrastructure]

The metricsets we used in the previous steps also comes with prebuilt dashboard

> that you can use to visualize the data. In Kibana, on the left panel, click on
> **dashboard**, search for EC2 and select the dashboard called: **[Metricbeat

Reviewer: ```suggestion
that you can use to visualize the data. In Kibana, open the main menu and click
**Dashboard**. Search for EC2 and select the dashboard called: **[Metricbeat
```
